js/logic.js

```diff
@@ -0,0 +1,862 @@
+
+    var prevPromise = opts.prevPromise || Promise.resolve();
+
+    if (this.not) {
+      this.promise = prevPromise.then(() => {
+        this.capturedLogs.some((event) => {
+          if ((!this.ns || event.namespace === this.ns) &&
+              event.matches(this.type, this.detailPredicate)) {
+            throw new MismatchError(this, event);
+          }
+        });
+      });
+    } else if (this.type) {
+      this.promise = new Promise((resolve, reject) => {
+        // Once any previous match has been resolved,
+        // subscribe to a following match.
+        prevPromise.then(() => {
+
+          var timeoutId = setTimeout(() => {
+            reject(new Error('LogicMatcherTimeout: ' + this));
+          }, this.timeoutMS);
+
+          // Promise chains have "dead spots" in between resolution
+          // callbacks. For instance:
+          //                 [promise1.then]      [promise2.then]
+          //    other events could be logged --^
+          //
+          // We could miss those events in the middle by just setting
+          // up a new listener for each LogicMatcher. Instead, since
+          // every matcher has a pointer to its prevMatcher, we can
+          // just grab the missing logs from there.
+          var removeMatchListener = (event) => {
+            this.resolved = true;
+            this.capturedLogs = []; // Extra events will go here.
+            setTimeout(function() {
+              emitter.removeListener('*', matchFn);
+            }, 0);
+          };
+
+          var matchFn = (event) => {
+            this.capturedLogs.push(event);
+            if (this.resolved) {
+              return;
+            }
+
+            if (this.ns && event.namespace !== this.ns ||
+                event.type !== this.type) {
+              return false; // did not match
+            }
+            if (event.matches(this.type, this.detailPredicate)) {
+              removeMatchListener(event);
+              this.matchedLogs.push(event);
+              clearTimeout(timeoutId);
+              logic(this, 'match', { ns: this.ns,
+                                     type: this.type,
+                                     event: event });
+              resolve(event);
+              return true;
+            } else {
+              if (this.failOnMismatchedDetails) {
+                removeMatchListener(event);
+                reject(new MismatchError(this, event));
+                return true; // matched
+              } else {
+                // Ignore mismatched events; maybe we'll match later.
+              }
+            }
+            return false; // not done yet, didn't find a match
+          };
+
+          emitter.on('*', matchFn);
+
+          if (opts.prevMatcher) {
+            // Run the match function on missed logs, until one of
+            // them returns true (indicating we matched and are done).
+            opts.prevMatcher.capturedLogs.some(matchFn);
```

Comment:
Logically, I think you want `forEach` here, not `some`.  Specifically, I think we want to propagate all of the logged events after the matching event forward for our child matchers.  Of course, forEach could result in some performance sadness, so the better thing I think would be to do something like:
```js
var prevLogs = opts.prevMatcher.capturedLogs;
var matchIndex = prevLogs.findIndex(matchFn);
if (matchIndex !== -1) {
  this.capturedLogs = prevLogs.slice(matchIndex + 1);
}
```